// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: transcriber.proto

package transcriberv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on VoiceText with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *VoiceText) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on VoiceText with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in VoiceTextMultiError, or nil
// if none found.
func (m *VoiceText) ValidateAll() error {
	return m.validate(true)
}

func (m *VoiceText) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Data

	// no validation rules for Confidence

	// no validation rules for SpeakerLabel

	// no validation rules for Language

	if len(errors) > 0 {
		return VoiceTextMultiError(errors)
	}

	return nil
}

// VoiceTextMultiError is an error wrapping multiple validation errors returned
// by VoiceText.ValidateAll() if the designated constraints aren't met.
type VoiceTextMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m VoiceTextMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m VoiceTextMultiError) AllErrors() []error { return m }

// VoiceTextValidationError is the validation error returned by
// VoiceText.Validate if the designated constraints aren't met.
type VoiceTextValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e VoiceTextValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e VoiceTextValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e VoiceTextValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e VoiceTextValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e VoiceTextValidationError) ErrorName() string { return "VoiceTextValidationError" }

// Error satisfies the builtin error interface
func (e VoiceTextValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sVoiceText.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = VoiceTextValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = VoiceTextValidationError{}

// Validate checks the field values on TranscribeResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *TranscribeResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TranscribeResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TranscribeResponseMultiError, or nil if none found.
func (m *TranscribeResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *TranscribeResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetResults() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, TranscribeResponseValidationError{
						field:  fmt.Sprintf("Results[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, TranscribeResponseValidationError{
						field:  fmt.Sprintf("Results[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return TranscribeResponseValidationError{
					field:  fmt.Sprintf("Results[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for SpeechEventType

	if all {
		switch v := interface{}(m.GetSpeechEventOffset()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TranscribeResponseValidationError{
					field:  "SpeechEventOffset",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TranscribeResponseValidationError{
					field:  "SpeechEventOffset",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetSpeechEventOffset()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TranscribeResponseValidationError{
				field:  "SpeechEventOffset",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return TranscribeResponseMultiError(errors)
	}

	return nil
}

// TranscribeResponseMultiError is an error wrapping multiple validation errors
// returned by TranscribeResponse.ValidateAll() if the designated constraints
// aren't met.
type TranscribeResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TranscribeResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TranscribeResponseMultiError) AllErrors() []error { return m }

// TranscribeResponseValidationError is the validation error returned by
// TranscribeResponse.Validate if the designated constraints aren't met.
type TranscribeResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TranscribeResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TranscribeResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TranscribeResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TranscribeResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TranscribeResponseValidationError) ErrorName() string {
	return "TranscribeResponseValidationError"
}

// Error satisfies the builtin error interface
func (e TranscribeResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTranscribeResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TranscribeResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TranscribeResponseValidationError{}

// Validate checks the field values on TranscribeRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *TranscribeRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TranscribeRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TranscribeRequestMultiError, or nil if none found.
func (m *TranscribeRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *TranscribeRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Audio

	if len(errors) > 0 {
		return TranscribeRequestMultiError(errors)
	}

	return nil
}

// TranscribeRequestMultiError is an error wrapping multiple validation errors
// returned by TranscribeRequest.ValidateAll() if the designated constraints
// aren't met.
type TranscribeRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TranscribeRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TranscribeRequestMultiError) AllErrors() []error { return m }

// TranscribeRequestValidationError is the validation error returned by
// TranscribeRequest.Validate if the designated constraints aren't met.
type TranscribeRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TranscribeRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TranscribeRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TranscribeRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TranscribeRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TranscribeRequestValidationError) ErrorName() string {
	return "TranscribeRequestValidationError"
}

// Error satisfies the builtin error interface
func (e TranscribeRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTranscribeRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TranscribeRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TranscribeRequestValidationError{}
